### DISCUSSION TITLE
Electron Forge: stylex.create should never be called. It should be compiled away.

### DISCUSSION THREAD

**Discussion by danieleperilli** [facebook/stylex#302] [Q&A]
**The problem**
I receive the `stylex.create should never be called. It should be compiled away.` error when I try to execute an Electron app (built with Forge) using React with SyleX.

![CleanShot 2024-01-03 at 14 54 03](https://github.com/facebook/stylex/assets/9072177/dbbf1616-a92b-4761-bfd1-0833644deb8e)

**How to reproduce**

Steps to reproduce:
1. Create new Electron app using electron-forge
2. Install StyleX as explained in the Webpack example on your website
3. Create a style in any component
4. Debug and run

**Configuration**
Here is a sample component that uses StyleX:

```ts
import * as stylex from '@stylexjs/stylex';
import { ChangeEvent, forwardRef, useRef, useImperativeHandle } from 'react';

interface Props {
    accept?: string[];
    type?: string;
    onFileChosen?: (file: File) => void;
}

interface Methods {
    browse: () => void;
}
export type FileBrowserMethods = Methods;

const styles = stylex.create({
    base: {
        display: "none",
        visibility: "hidden",
    },
});

const FileBrowser = forwardRef<Methods, Props>(({ accept, onFileChosen }, ref) => {
    
    const fileInputRef = useRef<HTMLInputElement>(null);

    const handleFileChange = (event: ChangeEvent<HTMLInputElement>) => {
        const files = event.target.files;
        if (!files || files.length === 0)
            return;

        const file = files[0];

        // Callback
        if (onFileChosen)
            onFileChosen(file);
    };

    const browse = ()=> {
        fileInputRef.current?.dispatchEvent(new MouseEvent("click"));
    };
    useImperativeHandle(ref, () => ({ browse }));

    return (
        <input type="file" ref={fileInputRef} accept={accept?.join(",")} onChange={handleFileChange} {...stylex.props(styles.base)} />
    );
})

export default FileBrowser;
```

Here is the webpack configuration (it's a bit different from your example because of forge):
![CleanShot 2024-01-03 at 15 00 32](https://github.com/facebook/stylex/assets/9072177/817be061-0ffa-4745-be57-1c6b9b0495b1)

The `webpack.plugins.ts` contains the configuration of StyleX plugin:
```ts
const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");
const StylexPlugin = require("@stylexjs/webpack-plugin");

export const plugins = [
    
    new StylexPlugin({
        filename: "styles.[contenthash].css",
        // dev: argv.mode === 'development',
        runtimeInjection: false,
        classNamePrefix: "x",
        unstable_moduleResolution: {
            type: "commonJS",
            rootDir: __dirname,
        },
    }),
    new ForkTsCheckerWebpackPlugin({
        logger: "webpack-infrastructure",
    }),
];
```
Any idea of what's going wrong?

Thanks!

**Comment by nmn** [comment]
Not sure how Forge works, but there is likely an issue with the build set up. Clearly, the code is not getting compiled as expected.

Try using just the Babel plugin with `dev: true` and see if you can get it to work.

**Reply by danieleperilli** [reply]
Unfortunately, it doesn't work. I tried installing the babel-loader and configuring it in the webpack config like this:

```ts
import type { ModuleOptions } from 'webpack';

export const rules: Required<ModuleOptions>['rules'] = [
    // Add support for native node modules
    {
        // We're specifying native_modules in the test because the asset relocator loader generates a
        // "fake" .node file which is really a cjs file.
        test: /native_modules[/\\].+\.node$/,
        use: "node-loader",
    },
    {
        test: /[/\\]node_modules[/\\].+\.(m?js|node)$/,
        parser: { amd: false },
        use: {
            loader: "@vercel/webpack-asset-relocator-loader",
            options: {
                outputAssetBase: "native_modules",
            },
        },
    },
    {
        test: /\.tsx?$/,
        exclude: /(node_modules|\.webpack)/,
        use: [
            
            // This is needed by StyleX
            {
                loader: "babel-loader",
                options: {
                    rootMode: "upward",
                },
            },

            {
                loader: "ts-loader",
                options: {
                    transpileOnly: true,
                },
            }
        ],
    },
    {
        test: /\.css$/,
        use: [
            { loader: "style-loader" }, 
            { loader: "css-loader" }
        ],
    },
    {
        test: /\.s[ac]ss$/i,
        use: [
            { loader: "style-loader" }, 
            { loader: "css-loader" },
            { loader: "sass-loader" }
        ],
    },
    
];
```
And I added the `babel.config.js` to the root of the dir as you suggested.
Nothing to do, same error.

To let you play with it (if you want), I created a basic repo and readme with the steps I followed (without the babel config you suggested).
https://github.com/danieleperilli/electron-with-stylex
You can fork it and just play the debug to run the application in VSCode.
Thanks!

**Reply by nmn** [reply]
With `dev:true`, (And `runtimeInject:true`) StyleX is a simple JS transform. So you should be able to find a way to make that work.

Even if the rest of the build is broken, you won't get a CSS file. But if you configure Babel correctly, you won't get runtime errors.

You have more experience with forge here, so please try to figure out how to add any custom Babel plugin and make it work.

**Reply by danieleperilli** [reply]
Thanks, I just committed the changes you suggested. However, no luck.

**Reply by danieleperilli** [reply]
Sorry, I just saw your updated reply. Well, I'm not so expert on Forge, but I can ask their team: https://github.com/electron/forge/issues/3459

**Reply by nmn** [reply]
Try making any custom Babel plugin work. Doesn't have to be StyleX.

**Reply by danieleperilli** [reply]
I was able to run the app without runtime errors, as you said, by removing the ts-loader and using the babel-loader. The StyleX class is added to the component, but the problem now is that the css file is not generated.
![CleanShot 2024-01-04 at 13 39 42](https://github.com/facebook/stylex/assets/9072177/f2fd91f8-105d-4685-8eec-3378e0acefef)

Honestly, I'm stuck :D

**Reply by nmn** [reply]
This is an improvement. Now try passing `dev: true` to the Babel plugin as an option. This relies on runtime style injection, but hopefully things should at least work.

**Reply by danieleperilli** [reply]
Well, it was already with dev: true, but it doesn't seem to be able to inject the style, or I am doing something wrong.
![CleanShot 2024-01-05 at 09 57 16](https://github.com/facebook/stylex/assets/9072177/0fd42314-db03-4fac-b92d-971a343cf15a)

**Reply by nmn** [reply]
What version of StyleX are you using?

**Reply by danieleperilli** [reply]
0.4.1

**Reply by thisisadarshjais** [reply]
@nmn  Do we need to always enable runtimeInjection for dev mode or  is there any way as of now to use statically generated CSS files and not runtime injected CSS.? Since I can't figure out any way to create static css during dev server (building the app generate css file inside dist, but during local dev no css file is getting generated) 

USING : "@stylexjs/stylex": "^0.5.1",

Incase, we need to explicitly enable runtimeinjection, Is there any specific use-case for webpack mode (dev:boolean), does it create any difference?

**Reply by nmn** [reply]
> Do we need to always enable runtimeInjection for dev mode or is there any way as of now to use statically generated CSS files and not runtime injected CSS.?

No, it's only needed if you want hot-reloading.

We know that our Webpack integration has issues and I have an issue with all my learning about Webpack. We're working to make this better. In the meantime, we're also working on a CLI that will solve most problems by sidestepping existing build processes entirely.
